### DISCUSSION TITLE
RestTemplate vs RestClient vs WebClient vs Feign

### DISCUSSION THREAD

**Discussion by GIVEN53** [2024-woowacourse-study/level-interview#199] [Spring 뒷고기]
.

**Comment by alstn113** [comment]
### RestTemplate
- 동기식 HTTP 클라이언트: 호출이 완료될 때까지 블로킹, 따라서 간단하고 직관적이지만, 비동기 프로세스에는 적합하지 않음.
- 템플릿 메서드 제공: 사용하기 쉬운 메서드를 통해 HTTP 요청을 간편하게 보낼 수 있음.
- 다양한 HTTP 메서드 지원: GET, POST, PUT, DELETE 등을 쉽게 호출 가능.
- 예외 처리 기능: 다양한 예외 처리가 가능하지만, 최신 기능에 비해 유연성이 떨어짐.
- Spring 5 이후로는 WebClient 권장: 최신 기능과 비동기 지원을 위해 WebClient로의 전환이 추천됨.

### RestClient
- 동기식 HTTP 클라이언트: 최신 Spring 6에서 도입, 블로킹 호출 방식.
- 플루언트 API: 더 읽기 쉬운 코드와 유연한 요청 설정 가능.
- 고급 요청 설정: 헤더, 쿼리 파라미터 등을 손쉽게 설정할 수 있음.
- RestTemplate의 대안: 기존 RestTemplate 사용자에게 더 모던한 대안을 제공하지만, 비동기식 작업에는 여전히 적합하지 않음.

### WebClient
- 비동기식 및 반응형 클라이언트: 프로젝트 리액터 기반, 비동기 프로세스를 효율적으로 처리 가능.
- 동기 및 비동기 호출 지원: 유연한 사용이 가능하지만, 초기 학습 곡선이 있음.
- 고급 에러 처리: 다양한 상태 코드와 예외를 처리할 수 있음.
- 최신 기술: Spring 5에서 도입, 미래 지향적인 설계.

### Feign
- 선언적인 REST 클라이언트: 인터페이스로 REST API 호출을 정의, 코드의 간결성과 가독성이 높음.
- Spring Cloud와 통합: 마이크로서비스 환경에서 쉽게 사용할 수 있음.
- 자동 구현: 인터페이스를 통해 REST 호출을 자동으로 생성, 설정이 간편하지만, 복잡한 시나리오에서는 제약이 있을 수 있음.
- Netflix 오픈 소스 라이브러리: 신뢰성 있

**Comment by 3Juhwan** [comment]
**RestTemplate**
- 간단한 HTTP 통신에 사용, Spring 3.0
- 간편한 사용
- 멀티 쓰레딩
- 동기식, 블록킹: 대규모 시스템에서 성능 문제
- 모던하지 못한 코드

**WebClient**
- Spring 5에 도입
- 모던한 코드 작성
- 싱글 쓰레드
- 동기, 비동기 통신 지원
- WebClent만을 사용하기 위해 WebFlux 의존성을 받아야 함

**RestClient**
- Spring 6.1에 도입된 HTTP 클라이언트
- 동기식 지원
- RestTemplate보다 세밀한 설정 가능 
- 모던한 코드, Spring의 다른 구성 요소와 잘 통합

**Feign**
- 넷플릭스가 만든 라이브러리
- Spring Cloud 같이 MSA 환경에서 주로 사용
- MSA 관련 기능 제공(로드 밸런싱, 서킷 브레이커)
- 선언적인 REST 클라이언트: 인터페이스로 REST API 호출을 정의, 코드의 간결성과 가독성이 높음.

--- 

**RestTemplate**

- 동기식: 요청을 보내고 응답을 받을 때까지 대기하는 방식입니다.
- Spring Framework의 일부: Spring 3.0부터 제공되었으며, 주로 간단한 HTTP 통신에 사용됩니다.
- 간편한 사용법: HTTP 메서드를 제공하여 간단한 HTTP 요청을 수행할 수 있습니다.
- 다양한 설정 가능: 인터셉터, 메시지 컨버터, 에러 핸들링 등 다양한 설정을 지원합니다.
- 단점: 블로킹 형태로 동작하여 대규모 애플리케이션에서는 성능 문제가 발생할 수 있습니다. (Spring 5부터는 비권장)

**RestClient**

- Spring Framework 6.1에서 도입된 새로운 HTTP 클라이언트: RestTemplate의 현대적인 대안입니다.
- 동기식 및 리액티브 지원: 동기 및 비동기식(리액티브) API를 모두 지원합니다.
- 커스터마이제이션: RestTemplate보다 더 많은 커스터마이제이션 기능을 제공하여 세밀한 설정이 가능합니다.
- Spring의 다른 구성 요소와 통합: Spring의 다른 구성 요소와 잘 통합되며, 모던 Java 특징을 활용할 수 있습니다.

**WebClient**

- Spring WebFlux에서 제공하는 비동기 HTTP 클라이언트: Spring 5에서 도입되었습니다.
- 리액티브 프로그래밍 지원: 비동기 방식으로 HTTP 요청을 처리하며, 리액티브 스트림을 사용하여 데이터를 처리합니다.
- 고급 기능 제공: 필터, 에러 처리, 로깅 등의 고급 기능을 제공하여 복잡한 HTTP 통신을 지원합니다.
- Spring MVC 및 WebFlux에서 모두 사용 가능: 기존의 Spring MVC와 리액티브 웹 애플리케이션에서 모두 사용할 수 있습니다.
- Spring WebFlux 의존성을 모두 받아야 하는 문제: WebClent만을 사용하기 위해 WebFlux를 받는 것은 고민이 필요합니다. 

**Feign**

- 선언적 REST 클라이언트: 인터페이스와 애너테이션을 사용하여 RESTful 서비스를 손쉽게 사용할 수 있게 해줍니다.
- Spring Cloud에서 널리 사용: Spring Cloud 프로젝트에서 서비스 간 통신을 단순화하는 데 유용합니다.
- Ribbon 및 Hystrix 통합: 클라이언트 측 로드 밸런싱(Ribbon)과 서킷 브레이커(Hystrix) 기능을 지원하여 안정적인 서비스 호출을 보장합니다.
- 설정 및 사용이 간편: 간단한 애너테이션과 인터페이스 정의를 통해 REST API를 호출할 수 있습니다.
